packages/model-viewer/src/features/scene-graph/model.ts

```diff
@@ -13,42 +13,226 @@
  * limitations under the License.
  */
 
-import {MeshStandardMaterial} from 'three';
+import {Material as ThreeMaterial, Mesh, MeshStandardMaterial, Object3D} from 'three';
+import {GLTFReference} from 'three/examples/jsm/loaders/GLTFLoader';
 
-import {CorrelatedSceneGraph} from '../../three-components/gltf-instance/correlated-scene-graph.js';
+import {CorrelatedSceneGraph, GLTFElementToThreeObjectMap, ThreeObjectSet} from '../../three-components/gltf-instance/correlated-scene-graph.js';
+import {GLTF, GLTFElement, KHRMaterialsVariants, Node as GLTFNode, Primitive} from '../../three-components/gltf-instance/gltf-2.0.js';
 
 import {Model as ModelInterface} from './api.js';
-import {Material} from './material.js';
+import {$ensureLoaded, Material} from './material.js';
 
 
 
 const $materials = Symbol('materials');
+const $variantInfo = Symbol('variantInfo');
+const $mesh = Symbol('mesh');
+const $children = Symbol('children');
+const $hierarchy = Symbol('hierarchy');
+const $roots = Symbol('roots');
+const $primitives = Symbol('primitives');
+export const $loadVariant = Symbol('loadVariant');
+export const $correlatedSceneGraph = Symbol('correlatedSceneGraph');
+export const $ensureAllMaterialsLoaded = Symbol('ensureAllMaterialsLoaded');
+
+
+// Holds onto temporary scene context information needed to perform lazy loading
+// of a resource.
+export class LazyLoader {
+  gltf: GLTF;
+  gltfElementMap: GLTFElementToThreeObjectMap;
+  mapKey: GLTFElement;
+  doLazyLoad: () => Promise<{set: ThreeObjectSet, material: ThreeMaterial}>;
+  constructor(
+      gltf: GLTF, gltfElementMap: GLTFElementToThreeObjectMap,
+      mapKey: GLTFElement,
+      doLazyLoad:
+          () => Promise<{set: ThreeObjectSet, material: ThreeMaterial}>) {
+    this.gltf = gltf;
+    this.gltfElementMap = gltfElementMap;
+    this.mapKey = mapKey;
+    this.doLazyLoad = doLazyLoad;
+  }
+}
+
+// Defines the base level node methods and data.
+class Node {
+  name: string = '';
+  [$children] = new Array<Node>();
+  constructor(name: string) {
+    this.name = name;
+  }
+}
+
+// Represents a primitive in a glTF mesh.
+class PrimitiveNode extends Node {
+  private[$mesh]: Mesh;
+  // Maps glTF material index number to a material that this primitive supports.
+  private[$materials] = new Map<number, Material>();
```

Comment:
You would still need to keep a list of indices into the original list in order to track which materials belong to this primitive, so it would require a mapping of indices to materials to do the look-up, hence why I condensed the relationship into a map. Not apposed to creating a list of indices used to lookup a material in the 'master list' but I don't see how that is cleaner.